01-architecture.md

```diff
@@ -52,38 +52,55 @@ One key constraint is that actor code size/footprint should be kept as low as po
 
-**Actor public interface**
-
-There are several alternatives:
-
-1. The outer layer performs method dispatch, against an exported table of callable methods defined by the actor.
-   - A standard/universal calling convention exists for method identification (current numerical method ID, string-based method ID, or something else).
-   - Pros:
-     - The environment can be aware of special methods such as "upgrade".
-     - IPLD schemas can be exported and environment can do type/contract checks.
-     - Makes actor interface introspectable.
-   - Cons:
-     - VM likely needs to maintain a cache to avoid analysing the exports table every time an actor is invoked.
-     - Complicates actor evolution (if we decide to support code mutability), as environment needs to deal with interface changes.
-     - Potential incompatibility in dealing with foreign VMs that do internal dispatch (e.g. EVM).
-2. The actor exposes a single entrypoint (e.g. run), and performs internal method dispatch.
-    - Calling convention is opaque and VM-specific.
-    - Method number can be one of three:
-      - 0: constructor.
-      - 1: send.
-      - 2: upgrade.
-    - Pros:
-      - Supports VMs that rely on internal dispatch (e.g. EVM).
-      - Environment stays agnostic to actor evolution.
-      - Simpler; removes complexity from the environment.
-      - Higher flexibility to accommodate sophisticated method dispatch, e.g. structural pattern matching, logic versioning, low-level proxying, etc.
-    - Cons:
-      - Because the calling convention is VM-specific, in cross-contract calls, the caller needs to know the VM type of the callee so it can pack the call adequately.
-
-We lean towards (2), and this places some constraints on usability:
-
-- SDK should provide utilities to facilitate internal dispatch based on method number _and_ VM's calling convention.
-- SDK should provide functions to pack message calls to foreign actors.
+## Actor public interface
+
+We considered two main approaches. Hybrids of these two approaches were also evaluated.
+
+**Approach 1 is external method dispatch.** The actor exports a table of callable methods. The sandbox matches the message to an exported method, either through the `MethodNum` message field, or by interpreting the input parameters string against a calling convention. This is largely how the current VM works today.
+
+**Approach 2 is for actors to expose a single entrypoint and rely on internal method dispatch.** This awards maximal degrees of freedom and sophistication to support techniques such as structural pattern matching, efficient pass-through proxying, interception, and more.
+
+Because internal dispatch will rapidly become boilerplate, the FVM SDK should offer utilities.
+
+Architecturally speaking, Approach 2 is more aligned with the actor model as implemented in the industry (Erlang, Akka, etc.), in which actors interact through single inboxes where messages are deposited.
```

Comment:
Maybe I should clarify for my own understanding. I get that you mean calling convention, but that of course includes the interfaces. How close are you thinking about emulating the actor model? Is part of the actor model idea to pass in the actor's state / context as an argument to the behaviour (e.g. BEAM/OTP), or will state be accessible implicitly ("globally") like in the EVM or Wasm bytecode?

Reply:
to continue the handwaving (🤣), while Actor models are more architecturally pure, many script kiddies (myself included) like having a global dispatcher/loop handle the execution of a function - so all i have to do is start my thing (e.g. a python function in a flask app) and then something else (e.g. a web browser calling a service) triggers execution.

I don't want to get too into the implementation details, but ensuring that we can mimic this form of execution - where all I have to do is write a function and _something_ (an external library, event, time, etc) triggers my function is a model that I think will be easier for a lot of people to grok.

TO BE CLEAR, I'm not saying we move away from the Actor model, just that we should invest in something (syntactic sugar? something else?) which lets novice coders (again, like me!) just write the basic stuff without having to understand many more complex topics.

Reply:
Yes, one thing doesn't preclude the other. A single entrypoint from the architectural perspective, combining with sugared dispatch support from the SDK, would make the development experience enjoyable.

Reply:
@expede yep, I called it a "hybrid" only because the pre-execution state is usually not encapsulated/self-contained in a _single object_ (a consequence of the IPLD model; although there may be actor state schemas that don't use links, probably very simple actors). So technically most actors will need to perform side effects in the form of read IO, so while execution is logically pure, it is not physically pure. (At least that's the way I see it!)